library/Freckle/App/Test/Http.hs

```diff
@@ -0,0 +1,332 @@
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedRecordDot #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE NoFieldSelectors #-}
+
+-- | Implements stubbing of an HTTP request function
+module Freckle.App.Test.Http
+  ( -- $docs
+    httpStubbed
+
+    -- * Defining stubs
+  , HttpStub (..)
+  , httpStub
+  , httpStubUrl
+
+    -- * Stub modifiers
+  , labelL
+  , MatchRequest (..)
+  , matchL
+
+    -- * Response modifiers
+  , statusL
+  , headersL
+  , bodyL
+
+    -- * Response helpers
+  , json
+
+    -- * FileSystem stubs
+  , loadHttpStubsDirectory
+
+    -- * 'MonadHttp' instances
+
+    -- ** For use with @DerivingVia@
+  , HasHttpStubs (..)
+  , ReaderHttpStubs (..)
+
+    -- ** Concrete transformer
+  , HttpStubsT
+  , runHttpStubsT
+  ) where
+
+import Freckle.App.Prelude
+
+import Control.Lens (Lens', lens, view, (&), (.~), (<>~))
+import Control.Monad (filterM)
+import Control.Monad.Reader (runReaderT)
+import Data.Aeson (ToJSON, encode)
+import qualified Data.ByteString.Lazy as BSL
+import Data.List (stripPrefix)
+import Data.String (IsString (..))
+import Freckle.App.Http (MonadHttp (..))
+import Freckle.App.Test.Http.MatchRequest
+import Network.HTTP.Client (Request, Response)
+import qualified Network.HTTP.Client.Internal as HTTP
+import Network.HTTP.Types.Header (ResponseHeaders, hAccept, hContentType)
+import Network.HTTP.Types.Status (Status, status200)
+import System.Directory (doesFileExist)
+import System.FilePath (addTrailingPathSeparator)
+import System.FilePath.Glob (globDir1)
+
+-- | Respond to a 'Request' with the first 'HttpStub' to match
+--
+-- If no stubs match, 'error' is used. If you'd rather experience a 404, add
+-- a final stub for any request that does that:
+--
+-- @
+-- stubs :: ['HttpStub']
+-- stubs =
+--   [ -- ...
+--   , -- ...
+--   , 'httpStub' \"Anything\" 'MatchAnything'
+--       & 'statusL' .~ 'status404'
+--       & 'bodyL' .~ \"Not found\"
+--   ]
+-- @
+httpStubbed
+  :: HasCallStack
+  => [HttpStub]
+  -> Request
+  -> Response BSL.ByteString
+httpStubbed stubs req =
+  maybe (error errorMessage) (toResponse req) $ headMay matched
+ where
+  (unmatched, matched) =
+    partitionEithers $
+      map
+        ( \stub ->
+            bimap (stub,) (const stub.response) $
+              matchRequest req stub.match
+        )
+        stubs
+
+  errorMessage =
+    "No stubs were found that matched:\n"
+      <> show req
+      <> "\n"
+      <> concatMap (uncurry unmatchedMessage) unmatched
+
+  unmatchedMessage stub err = "\n== " <> stub.label <> " ==\n" <> err
+
+-- | Fields that can be defined for a response
+data HttpStubResponse = HttpStubResponse
+  { status :: Status
+  , headers :: ResponseHeaders
```

Comment:
What do you think about abstracting this slightly further as `headers :: Map (CI ByteString) ByteString`?

Reply:
Well, I think it's generally best to match the type of the same field in `Response`, such to ensure we can represent any possible `Response` in the wild.

I had to think about this case in particular, and it does happen to matter. Not only does HTTP allow multiple values at the same header name, but there are actual semantics for it too: https://stackoverflow.com/a/4371395.

Reply:
The semantics for it are exactly why it seems reasonable to say we should be able to normalize responses that do have multiple fields of the same name into a standard form that doesn't.

Reply:
I think you're implying that normalizing the representation is better than just remaining faithful to the representation originally stubbed. Is that right? If so, why?

Reply:
Haven't really thought through why, it's just it seems like the purpose of this type already is to be a slightly higher-level representation of a response, as it doesn't represent the http version, exotic bits like trailer fields, etc. so thought I'd throw it out there.

Reply:
:shrug: